### PR TITLE
[CELEBORN-302] Fix workers count out of sync in HA mode.

### PR DESCRIPTION
**Pull request by FMX**
### What changes were proposed in this pull request?
The worker's list might add two identical workers if the masters failed to connect to the worker while setting up the endpoint.

### Why are the changes needed?
Change the worker's list to set to remove redundant workers.

### Does this PR introduce _any_ user-facing change?
NO.

### How was this patch tested?
UT.

### CODE REVIEW

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java [hunk 1]
```diff
@@ -51,7 +47,7 @@ public abstract class AbstractMetaManager implements IMetadataHandler {
   // Meta data for master service
   public final Set<String> registeredShuffle = ConcurrentHashMap.newKeySet();
   public final Set<String> hostnameSet = ConcurrentHashMap.newKeySet();
-  public final ArrayList<WorkerInfo> workers = new ArrayList<>();
+  public final Set<WorkerInfo> workers = new HashSet<>();
```
Comment (pan3793) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java:50]: The implementation of `WorkerInfo#equals` is fragile, e.g. what if the `obj` is not a `WorkerInfo` object?

Comment (FMX) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java:50]: Updated. How about now?

File: common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala [hunk 2]
```diff
@@ -259,17 +259,22 @@ class WorkerInfo(
...
-    host == other.host &&
-    rpcPort == other.rpcPort &&
-    pushPort == other.pushPort &&
-    fetchPort == other.fetchPort &&
-    replicatePort == other.replicatePort
+  private def canEqual(other: Any): Boolean = other.isInstanceOf[WorkerInfo]
+
+  override def equals(other: Any): Boolean = other match {
+    case that: WorkerInfo =>
+      that.canEqual(this) &&
```
Comment (waitinfuture) [inline common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala:266]: useless

File: common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala [hunk 3]
```diff
@@ -259,17 +259,22 @@ class WorkerInfo(
...
   }
 
-  override def equals(obj: Any): Boolean = {
-    val other = obj.asInstanceOf[WorkerInfo]
-    host == other.host &&
-    rpcPort == other.rpcPort &&
-    pushPort == other.pushPort &&
-    fetchPort == other.fetchPort &&
-    replicatePort == other.replicatePort
+  private def canEqual(other: Any): Boolean = other.isInstanceOf[WorkerInfo]
```
Comment (waitinfuture) [inline common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala:262]: useless

**Review by waitinfuture — APPROVED** [review]
LG TM, thanks!

**Pull request merged**